viewer/src/Main.rs

```diff
@@ -0,0 +1,324 @@
+use sdl2::event::Event;
+use sdl2::keyboard::Keycode;
+use sdl2::pixels::Color;
+use sdl2::rect::Rect;
+use sdl2::render::TextureCreator;
+use sdl2::render::WindowCanvas;
+use sdl2::ttf::Font;
+use sdl2::video::WindowContext;
+use std::env;
+use std::path::Path;
+use std::thread;
+use std::time::Duration;
+use std::io::prelude::*;
+use bytemuck::Pod;
+use bytemuck::Zeroable;
+use std::fs::File;
+
+#[derive(Clone, Copy, Debug, Pod, Zeroable)]
+#[repr(C)]
+pub struct Span {
+    tag: u64,
+    start: u64,
+    stop: u64,
+}
+
+pub struct Area {
+    y_start: u64,
+    x_start: i32,
+    y_stop: u64,
+    x_stop: i32,
+    tag_data: (u64,u64,u64),
+}
+
+pub struct App {
+    window_width: u32,
+    background_color: Color,
+    draw_zones: Vec<Area>,
+    colors: Vec<Color>,
+    muted_colors: Vec<Color>,
+    sdl_context: sdl2::Sdl,
+    canvas: WindowCanvas,
+    /// span cycles per horizontal pixel
+    scale: u64,
+    /// vertical pixels per span
+    span_height: i32,
+    /// vertical pixels between spans
+    span_spacing: i32,
+    min_start: u64,
+    max_stop: u64,
+    scroll: i32,
+    texture_creator: TextureCreator<WindowContext>,
+}
+
+impl App {
+    pub fn new(filled_spans: &mut Vec<Span>) -> Result<App, String> {
+        let mut spans: Vec<Span> = vec![];
+        for span in filled_spans{
+            spans.push(*span);
+        };
+        assert!(
+            !spans.is_empty(),
+            "expected a non-empty array of trace spans"
+        );
+        spans.sort_unstable_by_key(|s| s.start);
+        let min_start = spans[0].start;
+        // TODO check and correct this on first iteration
+        let max_stop = spans[spans.len() - 1].stop;
+        let draw_zones: Vec<Area> = vec![];
+        let window_width = 1600u32;
```

Comment:
I'm not sure what SDL does with windows that are larger than someone's screen size, will they still be able to get to the edge of the window to resize it?  Otherwise should probably default this to something safer, like 800.

If you end up introducing some kind of configuration file later, seems like a thing to include there.

Reply:
Talked about this behavior in person, will deal with it on a later PR.